test/parallel/test-readable-from.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+require('../common');
+const { once } = require('events');
+const { Readable } = require('stream');
+const { strictEqual } = require('assert');
+
+async function toReadableBasicSupport() {
+  async function * generate() {
+    yield 'a';
+    yield 'b';
+    yield 'c';
+  }
+
+  const stream = Readable.from(generate(), { objectMode: true });
+
+  const expected = ['a', 'b', 'c'];
+
+  for await (const chunk of stream) {
+    strictEqual(chunk, expected.shift());
+  }
+}
+
+async function toReadableOnData() {
+  async function * generate() {
+    yield 'a';
+    yield 'b';
+    yield 'c';
+  }
+
+  const stream = Readable.from(generate(), { objectMode: true });
+
+  let iterations = 0;
+  const expected = ['a', 'b', 'c'];
+
+  stream.on('data', (chunk) => {
+    iterations++;
+    strictEqual(chunk, expected.shift());
+  });
+
+  await once(stream, 'finish');
+
+  strictEqual(iterations, 3);
+}
+
+async function destroysTheStreamWhenThrowing() {
+  async function * generate() {
+    throw new Error('kaboom');
+  }
+
+  const stream = Readable.from(generate(), { objectMode: true });
+
+  stream.read();
+
+  try {
+    await once(stream, 'error');
+  } catch (err) {
+    strictEqual(err.message, 'kaboom');
+    strictEqual(stream.destroyed, true);
+  }
+}
+
+async function asTransformStream() {
+  async function * generate(stream) {
+    for await (const chunk of stream) {
+      yield chunk.toUpperCase();
+    }
+  }
+
+  const source = new Readable({
+    objectMode: true,
+    read() {
+      this.push('a');
+      this.push('b');
+      this.push('c');
+      this.push(null);
+    }
+  });
+
+  const stream = Readable.from(generate(source), { objectMode: true });
+
+  const expected = ['A', 'B', 'C'];
+
+  for await (const chunk of stream) {
+    strictEqual(chunk, expected.shift());
+  }
+}
+
+Promise.all([
```

Comment:
Please add a test for an async iterator that is not an async generator to these.

Reply:
Would also be interesting to add a test for an array of promises - would be cool if `Readable.from` explicitly supported this since `for await` does.

Reply:
> Would also be interesting to add a test for an array of promises - would be cool if `Readable.from` explicitly supported this since `for await` does.

This means explicitly support async iterables + plain iterables, in addition to just iterat*ors*, right? I like that.

Reply:
Really great suggestions! Instead of just supporting `.next()` I've rather explicitly implemented the Iterable protocols through `Symbol.asyncIterator` and `Symbol.iterator` equivalently.

In addition, in order to support the array of promises case it was necessary to refactor the processing to actually do an `await value`, so I refactored the internals to support this (previously this would have returned an explicit Promise even though its a `for await (chunk...`).

So now we support all these cases out of the box, and things like - `Readable.from('hello world')` running the string iterator all works. Appropriate tests included as well.